File: 2-hybrid-search-with-rrf-reranking/0-typescript/src/hybrid/hybrid-search.service.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { fuseByRrf } from './rrf';
import { FusedHit, UpstreamHit } from './types';

@Injectable()
export class HybridSearchService {
  private readonly bm25Url: string;
  private readonly vectorUrl: string;
  private readonly rrfK: number;

  constructor(config: ConfigService) {
    this.bm25Url = config.get<string>('app.bm25Url')!;
    this.vectorUrl = config.get<string>('app.vectorUrl')!;
    this.rrfK = config.get<number>('app.rrfK', 60);
  }

  /**
   * Fan out to both engines IN PARALLEL: total latency ~= max(bm25, vector), not
   * the sum. Each call has its own .catch so one failure cannot reject the whole
   * Promise.all — a failed upstream degrades to an empty list and RRF fuses whatever
   * survives instead of crashing the request.
   */
  async search(q: string, limit: number): Promise<FusedHit[]> {
    const fetchSize = Math.max(limit * 2, limit); // over-fetch so consensus docs are not truncated early.
    const [bm25List, vectorList] = await Promise.all([
      this.fetchRanked(this.bm25Url, q, fetchSize).catch(() => [] as UpstreamHit[]),
      this.fetchRanked(this.vectorUrl, q, fetchSize).catch(() => [] as UpstreamHit[]),
    ]);
    return fuseByRrf(bm25List, vectorList, this.rrfK, limit);
  }

  /** Pass-through to a single upstream; returns its untouched ranked list. */
  async passthrough(url: string, q: string, limit: number): Promise<UpstreamHit[]> {
    return this.fetchRanked(url, q, limit);
  }

  get bm25(): string {
    return this.bm25Url;
  }

  get vector(): string {
    return this.vectorUrl;
  }

  private async fetchRanked(
    url: string,
    q: string,
    limit: number,
  ): Promise<UpstreamHit[]> {
    const res = await fetch(
      `${url}?q=${encodeURIComponent(q)}&limit=${limit}`,
    );
    if (!res.ok) throw new Error(`upstream ${url} returned ${res.status}`);
    const body = (await res.json()) as { hits: UpstreamHit[] };
    return body.hits ?? [];
  }
}

File: 2-hybrid-search-with-rrf-reranking/1-java/src/main/java/com/starci/hybrid/HybridSearchService.java
package com.starci.hybrid;

import com.starci.hybrid.Models.Doc;
import com.starci.hybrid.Models.Hit;
import com.starci.hybrid.Models.UpstreamResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.List;

@Service
public class HybridSearchService {

    private final WebClient bm25Client;
    private final WebClient vectorClient;
    private final int rrfK;

    public HybridSearchService(WebClient.Builder builder,
                               @Value("${upstream.bm25Url:http://bm25-service:4001/search}") String bm25Url,
                               @Value("${upstream.vectorUrl:http://vector-service:4002/search}") String vectorUrl,
                               @Value("${rrf.k:60}") int rrfK) {
        // One non-blocking WebClient per upstream; the connection pool is shared.
        this.bm25Client = builder.baseUrl(bm25Url).build();
        this.vectorClient = builder.baseUrl(vectorUrl).build();
        this.rrfK = rrfK;
    }

    /**
     * Fan out to both upstreams in parallel via Mono.zip. Each branch degrades to an
     * empty list on error (failure isolation), so one downed upstream does not fail
     * the whole request. Total latency is bounded by the slower upstream.
     */
    public Mono<List<Hit>> search(String q, int limit) {
        int fetchSize = Math.max(limit * 2, limit);
        Mono<List<Doc>> bm25 = fetch(bm25Client, q, fetchSize).onErrorReturn(List.of());
        Mono<List<Doc>> vector = fetch(vectorClient, q, fetchSize).onErrorReturn(List.of());
        return Mono.zip(bm25, vector)
                .map(t -> Rrf.fuse(t.getT1(), t.getT2(), rrfK, limit));
    }

    /** Pass-through to a single upstream; returns its untouched ranked list. */
    public Mono<List<Doc>> passthrough(boolean isBm25, String q, int limit) {
        return fetch(isBm25 ? bm25Client : vectorClient, q, limit).onErrorReturn(List.of());
    }

    private Mono<List<Doc>> fetch(WebClient client, String q, int limit) {
        return client.get()
                .uri(uri -> uri.queryParam("q", q).queryParam("limit", limit).build())
                .retrieve()
                .bodyToMono(UpstreamResponse.class)
                .map(UpstreamResponse::hits);
    }
}
